### PR TITLE
[wip] Add bucket sns notification support to attach-encrypt

### LINKED ISSUE #297: attach-encrypt should support bucket sns notification

### PR DESCRIPTION
**Pull request by whit537**
Closes #297. Includes #869.

### CODE REVIEW

File: c7n/mu.py [hunk 1]
```diff
@@ -991,6 +991,82 @@ def remove(self, func):
...
+                    try:
+                        response = self.client.unsubscribe(
+                            SubscriptionArn=subscription['SubscriptionArn'])
+                        log.debug("Unsubscribed %s from %s" %
+                            (func.name, topic_name))
+                    except ClientError as e:
+                        code = e.response['Error']['Code']
+                        if code != 'ResourceNotFoundException':
+                            raise
+                    break
```
Comment (kapilt) [inline c7n/mu.py]: minor, fwiw, you actually need to break twice to get out of the double for loop.

**Comment by whit537** [comment]
Actually, it seems that `attach-encrypt` [doesn't use](https://github.com/capitalone/cloud-custodian/blob/203fd33931442903fdf98373e025b000c6aea1f0/c7n/resources/s3.py#L623-L625) `s3crypt`'s `buckets` param, but rather [adds a source](https://github.com/capitalone/cloud-custodian/blob/203fd33931442903fdf98373e025b000c6aea1f0/c7n/resources/s3.py#L668-L672) in `process_buckets`.

File: c7n/mu.py (hunk 1 above)
Comment (whit537) [inline c7n/mu.py]: Addressed in 8f2915aec1e35945e486c944bde2bc89a1d8f70a over on #869.

**Comment by whit537** [comment]
From https://github.com/capitalone/cloud-custodian/issues/297#issuecomment-277978852:

> it provisioned a lambda function that will encrypt objects as they are added to the bucket. it did so by setting a bucket notification. extant buckets may already have a bucket notification in which case this won't work, per s3 bucket notification capabilities. @ocampocj i believe was looking in the case of an sns notification on a bucket to attach the lambda to the topic.

Okay, so my plan here is to:

- [ ] Get `attach-encrypt` working with the existing bucket notification functionality.
- [ ] Get it working with SNS instead.

Verbose logging on https://github.com/whit537/cloud-custodian/issues/1 per https://github.com/capitalone/cloud-custodian/issues/297#issuecomment-277978378.

**Comment by whit537** [comment]
Rebased to pick up #869, was 203fd33931442903fdf98373e025b000c6aea1f0.

**Pull request closed without merging**